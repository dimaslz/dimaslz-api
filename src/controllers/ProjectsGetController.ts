import { Request, Response } from "express";
import httpStatus from "http-status";

import { Controller } from "./Controller";

export default class ProjectsGetController implements Controller {
	async run(req: Request<{ attribute: string }, never, never, never>, res: Response) {
		let {
			default: { projects },
		}: any = await import("../data");

		const { s } = req.query;

		if (Array.isArray(projects) && s) {
			projects = projects.filter((item) => {
				return item.keywords.some((keyword) => keyword.includes(s));
			});
		}

		res.status(httpStatus.OK);

		return res.json({
			data: projects,
		});
	}
}
