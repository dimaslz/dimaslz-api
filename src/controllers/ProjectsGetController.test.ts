import httpStatus from "http-status";

import ProjectsGetController from "./ProjectsGetController";

describe("Projects get controller: GET /projects", () => {
	it("should return all projects by default", async () => {
		const projectsGetController: ProjectsGetController = new ProjectsGetController();

		const status = jest.fn((v) => v);
		const json = jest.fn((v) => v);
		const result = await projectsGetController.run.bind(projectsGetController)(
			{
				query: { s: null },
			},
			{
				status,
				json,
			}
		);

		expect(status).toHaveBeenCalledWith(httpStatus.OK);
		expect(result.data.length).toBe(9);

		expect(result).toMatchSnapshot();
	});

	it("should return projects which match with the query", async () => {
		const projectsGetController: ProjectsGetController = new ProjectsGetController();

		const status = jest.fn((v) => v);
		const json = jest.fn((v) => v);
		const result = await projectsGetController.run.bind(projectsGetController)(
			{
				query: { s: "typescript" },
			},
			{
				status,
				json,
			}
		);

		expect(status).toHaveBeenCalledWith(httpStatus.OK);
		expect(result.data.length).toBe(7);

		expect(result).toMatchSnapshot();
	});
});
