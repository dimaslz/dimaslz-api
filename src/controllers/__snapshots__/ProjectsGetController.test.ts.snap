// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects get controller: GET /projects should return all projects by default 1`] = `
Object {
  "data": Array [
    Object {
      "description": "An Angular components library to use Heroicons.com in your Angular projects.",
      "isOpenSource": true,
      "keywords": Array [
        "typescript",
        "javascript",
        "angular",
        "nodejs",
        "library",
        "components",
        "tailwindcss",
        "npm",
      ],
      "name": "ng-heroicons",
      "npm": "https://www.npmjs.com/package/@dimaslz/ng-heroicons",
      "repository": Array [
        "https://github.com/dimaslz/ng-heroicons",
      ],
      "stack": Array [
        "typescript",
        "javascript",
        "angular",
        "nodejs",
        "tailwindcss",
      ],
      "url": "https://ng-heroicons.dimaslz.dev/",
    },
    Object {
      "description": "A tool to create mock Api responses with your custom schema.",
      "isOpenSource": false,
      "keywords": Array [
        "typescript",
        "preact",
        "nodejs",
        "vitejs",
        "tools",
        "api",
        "tailwindcss",
      ],
      "name": "Randomdata",
      "npm": null,
      "repository": Array [],
      "stack": Array [
        "typescript",
        "preact",
        "nodejs",
        "vitejs",
        "tailwindcss",
      ],
      "url": "https://randomdata.loremapi.io/",
    },
    Object {
      "description": "A tool to create a framework icon component from a SVG",
      "isOpenSource": false,
      "keywords": Array [
        "nextjs",
        "nodejs",
        "tailwindcss",
        "tools",
      ],
      "name": "Framework icon builder",
      "npm": null,
      "repository": Array [],
      "stack": Array [
        "nextjs",
        "nodejs",
        "tailwindcss",
      ],
    },
    Object {
      "description": "Deploy any project easily, like Vercel or Heroku but in your self hosting.",
      "isOpenSource": false,
      "keywords": Array [
        "vitejs",
        "typescript",
        "vue3",
        "pinia",
        "tailwindcss",
        "docker",
        "tools",
        "github actions",
      ],
      "name": "Fito deploy",
      "npm": null,
      "repository": Array [],
      "stack": Array [
        "vitejs",
        "typescript",
        "vue3",
        "pinia",
        "tailwindcss",
        "docker",
        "github actions",
      ],
      "url": "https://fito-deploy.dimaslz.dev",
    },
    Object {
      "description": "Mock and document your Api's",
      "isOpenSource": false,
      "keywords": Array [
        "angular",
        "nodejs",
        "tailwindcss",
        "typescript",
        "tools",
        "api",
        "developers",
      ],
      "name": "Loremapi",
      "npm": null,
      "repository": Array [],
      "stack": Array [
        "angular",
        "nodejs",
        "tailwindcss",
        "typescript",
      ],
      "url": "https://loremapi.io",
    },
    Object {
      "description": "My online CV",
      "isOpenSource": true,
      "keywords": Array [
        "svelte",
        "tailwindcss",
        "typescript",
        "cv",
        "landing",
      ],
      "name": "cv.dimaslz.dev",
      "npm": null,
      "repository": Array [
        "https://github.com/dimaslz/dimaslz-cv",
      ],
      "stack": Array [
        "svelte",
        "tailwindcss",
        "typescript",
      ],
      "url": "https://cv.dimaslz.dev",
    },
    Object {
      "description": "My professional info by API",
      "isOpenSource": true,
      "keywords": Array [
        "nodejs",
        "typescript",
        "cv",
        "api",
      ],
      "name": "api.dimaslz.dev",
      "npm": null,
      "repository": Array [
        "https://github.com/dimaslz/dimaslz-api",
      ],
      "stack": Array [
        "nodejs",
        "typescript",
      ],
      "url": "https://api.dimaslz.dev",
    },
    Object {
      "description": "Dev landing",
      "isOpenSource": true,
      "keywords": Array [
        "javascript",
        "tailwindcss",
        "landing",
      ],
      "name": "dimaslz.dev",
      "npm": null,
      "repository": Array [
        "https://github.com/dimaslz/dimaslz.io-landing",
      ],
      "stack": Array [
        "javascript",
        "tailwindcss",
      ],
      "url": "https://dimaslz.dev",
    },
    Object {
      "description": "Profesional landing profile",
      "isOpenSource": true,
      "keywords": Array [
        "nextjs",
        "typescript",
        "tailwindcss",
      ],
      "name": "dimaslz.com",
      "npm": null,
      "repository": Array [
        "https://github.com/dimaslz/dimaslz.io",
      ],
      "stack": Array [
        "nextjs",
        "typescript",
        "tailwindcss",
      ],
      "url": "https://dimaslz.com",
    },
  ],
}
`;

exports[`Projects get controller: GET /projects should return projects which match with the query 1`] = `
Object {
  "data": Array [
    Object {
      "description": "An Angular components library to use Heroicons.com in your Angular projects.",
      "isOpenSource": true,
      "keywords": Array [
        "typescript",
        "javascript",
        "angular",
        "nodejs",
        "library",
        "components",
        "tailwindcss",
        "npm",
      ],
      "name": "ng-heroicons",
      "npm": "https://www.npmjs.com/package/@dimaslz/ng-heroicons",
      "repository": Array [
        "https://github.com/dimaslz/ng-heroicons",
      ],
      "stack": Array [
        "typescript",
        "javascript",
        "angular",
        "nodejs",
        "tailwindcss",
      ],
      "url": "https://ng-heroicons.dimaslz.dev/",
    },
    Object {
      "description": "A tool to create mock Api responses with your custom schema.",
      "isOpenSource": false,
      "keywords": Array [
        "typescript",
        "preact",
        "nodejs",
        "vitejs",
        "tools",
        "api",
        "tailwindcss",
      ],
      "name": "Randomdata",
      "npm": null,
      "repository": Array [],
      "stack": Array [
        "typescript",
        "preact",
        "nodejs",
        "vitejs",
        "tailwindcss",
      ],
      "url": "https://randomdata.loremapi.io/",
    },
    Object {
      "description": "Deploy any project easily, like Vercel or Heroku but in your self hosting.",
      "isOpenSource": false,
      "keywords": Array [
        "vitejs",
        "typescript",
        "vue3",
        "pinia",
        "tailwindcss",
        "docker",
        "tools",
        "github actions",
      ],
      "name": "Fito deploy",
      "npm": null,
      "repository": Array [],
      "stack": Array [
        "vitejs",
        "typescript",
        "vue3",
        "pinia",
        "tailwindcss",
        "docker",
        "github actions",
      ],
      "url": "https://fito-deploy.dimaslz.dev",
    },
    Object {
      "description": "Mock and document your Api's",
      "isOpenSource": false,
      "keywords": Array [
        "angular",
        "nodejs",
        "tailwindcss",
        "typescript",
        "tools",
        "api",
        "developers",
      ],
      "name": "Loremapi",
      "npm": null,
      "repository": Array [],
      "stack": Array [
        "angular",
        "nodejs",
        "tailwindcss",
        "typescript",
      ],
      "url": "https://loremapi.io",
    },
    Object {
      "description": "My online CV",
      "isOpenSource": true,
      "keywords": Array [
        "svelte",
        "tailwindcss",
        "typescript",
        "cv",
        "landing",
      ],
      "name": "cv.dimaslz.dev",
      "npm": null,
      "repository": Array [
        "https://github.com/dimaslz/dimaslz-cv",
      ],
      "stack": Array [
        "svelte",
        "tailwindcss",
        "typescript",
      ],
      "url": "https://cv.dimaslz.dev",
    },
    Object {
      "description": "My professional info by API",
      "isOpenSource": true,
      "keywords": Array [
        "nodejs",
        "typescript",
        "cv",
        "api",
      ],
      "name": "api.dimaslz.dev",
      "npm": null,
      "repository": Array [
        "https://github.com/dimaslz/dimaslz-api",
      ],
      "stack": Array [
        "nodejs",
        "typescript",
      ],
      "url": "https://api.dimaslz.dev",
    },
    Object {
      "description": "Profesional landing profile",
      "isOpenSource": true,
      "keywords": Array [
        "nextjs",
        "typescript",
        "tailwindcss",
      ],
      "name": "dimaslz.com",
      "npm": null,
      "repository": Array [
        "https://github.com/dimaslz/dimaslz.io",
      ],
      "stack": Array [
        "nextjs",
        "typescript",
        "tailwindcss",
      ],
      "url": "https://dimaslz.com",
    },
  ],
}
`;
